### PR TITLE
Wait for background JumpDestinationAnalysis if in progress

### PR DESCRIPTION
**Pull request by benaadams**
## Changes

- If its already started processing it will likely finish before starting the processing from start will
- Also replaced queued/received section in logs which are always 0, with from cache + analysed which is more interesting data

image

#### What types of changes does your code introduce?

- [x] Optimization

#### Requires testing

- [x] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Evm/CodeAnalysis/JumpDestinationAnalyzer.cs [hunk 1]
```diff
@@ -19,17 +20,56 @@ public sealed class JumpDestinationAnalyzer(ReadOnlyMemory<byte> code)
...
+                Thread thread = Thread.CurrentThread;
+                ThreadPriority priority = thread.Priority;
+                // We are waiting, so drop priority to normal (BlockProcessing runs at higher priority).
+                thread.Priority = ThreadPriority.Normal;
+
+                // Already in progress, wait for completion.
+                resetEvent.Wait();
+
+                // Restore the priority of the thread.
+                thread.Priority = priority;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/CodeAnalysis/JumpDestinationAnalyzer.cs:67]: should this be in finally?

File: src/Nethermind/Nethermind.Evm/CodeAnalysis/JumpDestinationAnalyzer.cs [hunk 2]
```diff
@@ -164,8 +207,28 @@ private static void MarkJumpDestinations(long[] jumpDestinationBitmap, int pos,
...
+                    thread.Priority = ThreadPriority.AboveNormal;
+
+                    _jumpDestinationBitmap ??= CreateJumpDestinationBitmap();
+
+                    // Release the MRES to be GC'd
+                    _analysisComplete = _jumpDestinationBitmap;
+                    // Signal complete.
+                    analysisComplete.Set();
+                    // Restore the priority of the thread.
+                    thread.Priority = priority;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/CodeAnalysis/JumpDestinationAnalyzer.cs:227]: should this be in finally?

**Review by LukaszRozmej — COMMENTED** [review]
The queues in logs are very useful in full sync or catching up to head.

File: src/Nethermind/Nethermind.Evm/CodeAnalysis/JumpDestinationAnalyzer.cs (hunk 1 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Evm/CodeAnalysis/JumpDestinationAnalyzer.cs:67]: This on definately; other not so important as threadpool resets it anyway; however better if we do as that's implementation detail

**Comment by benaadams** [comment]
> The queues in logs are very useful in full sync or catching up to head.

Switch modes when the queues are empty?

image

**Comment by benaadams** [comment]
Though not happy with `scontracts` as a term 😅

**Comment by benaadams** [comment]
`exec code  from cache     982 | new     11` instead

image

File: src/Nethermind/Nethermind.Consensus/Processing/ProcessingStats.cs [hunk 3]
```diff
@@ -200,8 +202,22 @@ public void UpdateStats(Block? block, IBlockTree blockTreeCtx, int recoveryQueue
...
+
+                    string extraStats;
+                    if (recoveryQueue > 0 || processingQueue > 0)
+                    {
+                        extraStats = $"recover {recoveryQueue,5:N0} | process {processingQueue,5:N0}";
+                    }
+                    else
+                    {
+                        extraStats = $"exec code {resetColor} from cache {cachedContractsUsed,7:N0} |{resetColor} new {contractsAnalysed,6:N0}";
+                    }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Processing/ProcessingStats.cs:217]: Hm... previously we created one string in `UpdateStats`, now we are creating 3.
I would in-line this into `_logger.Info()` call.

File: src/Nethermind/Nethermind.Evm/CodeAnalysis/JumpDestinationAnalyzer.cs [hunk 4]
```diff
@@ -19,17 +20,62 @@ public sealed class JumpDestinationAnalyzer(ReadOnlyMemory<byte> code)
...
+            if (previous is null)
+            {
+                ManualResetEventSlim analysisComplete = new(initialState: false);
+                previous = Interlocked.CompareExchange(ref _analysisComplete, analysisComplete, null);
+                if (previous is null)
+                {
+                    // Not already in progress, so start it.
+                    var bitmap = CreateJumpDestinationBitmap();
+                    _jumpDestinationBitmap = bitmap;
+                    // Release the MRES to be GC'd
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/CodeAnalysis/JumpDestinationAnalyzer.cs:48]: Would be good to Dispose MRES too.

File: src/Nethermind/Nethermind.Evm/CodeAnalysis/JumpDestinationAnalyzer.cs [hunk 5]
```diff
@@ -19,17 +20,62 @@ public sealed class JumpDestinationAnalyzer(ReadOnlyMemory<byte> code)
...
+                    _analysisComplete = bitmap;
+                    // Signal complete.
+                    analysisComplete.Set();
+                    return bitmap;
+                }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/CodeAnalysis/JumpDestinationAnalyzer.cs:53]: Dispose it too when set fails

File: src/Nethermind/Nethermind.Evm/CodeAnalysis/JumpDestinationAnalyzer.cs [hunk 6]
```diff
@@ -164,8 +213,34 @@ private static void MarkJumpDestinations(long[] jumpDestinationBitmap, int pos,
...
+                    ThreadPriority priority = thread.Priority;
+                    try
+                    {
+                        // Boost the priority of the thread as block processing may be waiting on this.
+                        thread.Priority = ThreadPriority.AboveNormal;
+
+                        _jumpDestinationBitmap ??= CreateJumpDestinationBitmap();
+
+                        // Release the MRES to be GC'd
+                        _analysisComplete = _jumpDestinationBitmap;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/CodeAnalysis/JumpDestinationAnalyzer.cs:231]: Dispose MRES?

File: src/Nethermind/Nethermind.Consensus/Processing/ProcessingStats.cs (hunk 3 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Consensus/Processing/ProcessingStats.cs:217]: Should we update stats outside the main thread?

Comment (benaadams) [reply src/Nethermind/Nethermind.Consensus/Processing/ProcessingStats.cs]: Moved to background thread. Doesn't use StringBuilder? Is 3 string per second isn't much in the scheme of things?

File: src/Nethermind/Nethermind.Evm/CodeAnalysis/JumpDestinationAnalyzer.cs (hunk 6 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Evm/CodeAnalysis/JumpDestinationAnalyzer.cs:231]: MRES doesn't have a finalizer and is only when you access the WaitHandle property (as we do with `WaitHandleExtensions`) that it creates a finalizable object that needs to be disposed. 

Regular MRE needs to be disposed; but MRES doesn't use any OS sync primitives unless you access the WaitHandle and cause it to allocate one.

File: src/Nethermind/Nethermind.Consensus/Processing/ProcessingStats.cs (hunk 3 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Consensus/Processing/ProcessingStats.cs]: Generating on background thread looks a good call

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Consensus/Processing/ProcessingStats.cs]: But each string interpolation creates a StringBuilder?

Comment (benaadams) [reply src/Nethermind/Nethermind.Consensus/Processing/ProcessingStats.cs]: Post C#10 .NET6.0 is complicated https://devblogs.microsoft.com/dotnet/string-interpolation-in-c-10-and-net-6/
image

And it caches via ArrayPool withthe main allocations happening post Log.Info in NLogger

image

**Comment by benaadams** [comment]
Added odd/even for peers report as its frequent enough that can see both and too frequent to have both at same time (two lines)

image

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**